### ISSUE TITLE
Type piracy in package

### ISSUE THREAD

**Issue by KristofferC** [JuliaLang/StyledStrings.jl#61]
All of this stuff:

https://github.com/JuliaLang/StyledStrings.jl/blob/6901610ac7809d8fb492a8ff588667cdd3f28862/src/io.jl#L252-L297

looks to be a pretty bad type piracy and should probably be moved to Base.

**Comment by kimikage** [comment]
dup of #57?

**Comment by tecosaur** [comment]
Loosely at best, since here we're talking about moving some functionality from this library to Base, and #57 is more about what/how functionality should be backported to the compat version.

**Comment by kimikage** [comment]
That's right. However, isn't the "result" of that the same?
That is, when they are transferred to `Base`, `julia1-compat` pirates them anew.

**Comment by tecosaur** [comment]
I can't say I follow. In Julia >1.11 the `julia1-compat` version shouldn't be used.

**Comment by kimikage** [comment]
I am just referring to `julia1-compat`, i.e., in Julia <1.11.

The above piracy could be eliminated from StyledString as *stdlib*. As a result, the piracy in `julia1-compat` will go from "mere backporting" to "pure piracy".

**Comment by ronisbr** [comment]
Hi!

Just for the records, it seems that this issue is causing a huge precompilation overhead in PrettyTables.jl when using Julia 1.11. For more information, see JuliaLang/julia#56080.

**Comment by tecosaur** [comment]
Good to have, and FWIW you have some sympathy for the effort in tracking this down.

I don't think the core issue has actually been stated here, so it would probably be worth me doing so for the record.

During the series of triage discussions that occurred around `StyledStrings`, it was decided that an initial proposal to add the functionality to `Base` would be split into a `Base` part and a new stdlib part (`StyledStrings`). This was done to prioritise keeping `Base` slim, but had the side-effect of creating the invalidations giving us headaches now.

There's ongoing discussion about how this should best be resolved, ranging from just evicting everything currently in `Base` (and the convenience that comes with it), to defining a "stub" `print` method in Base to control invalidations that's (somehow) substituted for the "fully-fledged" `print` defined in this library, or even brining all the code required for `print`ing an `AnnotatedString` with color into `Base`.

Cody also had an interesting idea I half-remember about making a more central, pluggable, print method to solve this issue and make it possible to customise printing potentially.

This is an issue that's cared about, without an obvious fix, but with a few ideas that need exploring.

**Comment by topolarity** [comment]
> Cody also had an interesting idea I half-remember about making a more central, pluggable, print method to solve this issue and make it possible to customise printing potentially.

Yeah the idea is to split the behavior into the `AnnotatedString` vs. `Face` portions, so that the AnnotatedString portion (which iterates annotations) can live in Base and the Face portion (which updates the actual printer state for display, resolves colors, etc.) can be separate

The additional wrinkle right now is that an `AnnotatedString` is just a set of Symbols mapped to regions, so we don't know which copy of StyledStrings in memory to ask to generate/display Faces for them (see https://github.com/JuliaLang/StyledStrings.jl/issues/91). We might need to associate an AnnotatedString with some kind of library-specific type that allows us to dispatch the resolution / display of its annotations to the correct library.

You wouldn't be allowed to "mix" annotations from different libraries in the same string - but that also dramatically simplifies the display logic since you only need to worry about "your own" display state on the library side. The key thing to get rid of is the hard-coded behavior where we assume there is one special library in memory that knows how to display annotations in an AnnotatedString (and we invalidate on it)

**Comment by KristofferC** [comment]
It seems to invalidate the package precompilation functionality:

```
❯ julia +nightly --project -e '@time @eval Base.Precompilation.precompilepkgs(); 
                                                  using StyledStrings; 
                                                   @time @eval Base.Precompilation.precompilepkgs()'
  0.082379 seconds (84.51 k allocations: 10.849 MiB, 51.15% compilation time)
  1.092273 seconds (1.46 M allocations: 83.460 MiB, 4.71% gc time, 96.65% compilation time: 100% of which was recompilation)
```

Causes latency problems in e.g Pkg and package loading when the REPL (which depends on StyledStrings) is loaded.

**Comment by tecosaur** [comment]
Will be closed by #131
